### ISSUE TITLE
error for 'vg surject'

### ISSUE THREAD

**Issue by wangl3** [vgteam/vg#4088]
**1. What were you trying to do?**

I first used 'vg giraffe' to map short reads to the reference, then used 'vg augment' to augment the graph. I am trying to use 'vg surject' to project the augmented alignment to BAM.
I used commands as below:
```
vg convert -t 4 ebv.d2.xg -p > ebv.pg
vg augment ebv.pg mapped.gam -m 10 -q 20 -Q 20 -S -A ebv-aug.gam > ebv-aug.pg 
vg convert -t 4 ebv-aug.pg -x > ebv-aug.xg 
vg surject -t 4 -b -x ebv-aug.xg ebv-aug.gam > ebv-aug.bam
```

**2. What did you want to happen?**
I am expecting it to generate BAM after augmentation.

**3. What actually happened?**
I didn't get BAM file but error messages as below:
```
error[Surjector::surject]: read M01853:259:000000000-KCLW9:1:1101:20339:8721 has 150 sequence bases but an input alignment that aligns 151 bases instead. This is invalid and uninterpretable; check your mapper.
error[Surjector::surject]: offending alignemnt: {"annotation": {"fragment_length": 327, "fragment_length_distribution": "-I 312.298947 -D 103.842894", "last_placed_stage": "none", "last_placed_stage_0bp": "none", "mapq_applied_cap": 244.56230356423899, "mapq_explored_cap": 0, "mapq_score_group": 2147483647, "mapq_uncapped": 2147483647, "proper_pair": true, "rescued": true, "secondary_scores": [266.99275582402436]}, "fragment_prev": {"name": "M01853:259:000000000-KCLW9:1:1101:20339:8721"}, "identity": 0.94039735099337751, "mapping_quality": 60, "name": "M01853:259:000000000-KCLW9:1:1101:20339:8721", "path": {"mapping": [{"edit": [{"from_length": 11, "to_length": 11}], "position": {"is_reverse": true, "node_id": "13769"}}, {"edit": [{"from_length": 1, "to_length": 1}], "position": {"is_reverse": true, "node_id": "14133"}}, {"edit": [{"from_length": 3, "to_length": 3}], "position": {"is_reverse": true, "node_id": "13768"}}, {"edit": [{"from_length": 1, "to_length": 1}], "position": {"is_reverse": true, "node_id": "13767"}}, {"edit": [{"from_length": 4, "to_length": 4}], "position": {"is_reverse": true, "node_id": "13766"}}, {"edit": [{"from_length": 113, "to_length": 113}], "position": {"is_reverse": true, "node_id": "13765"}}, {"edit": [{"from_length": 1, "to_length": 1}], "position": {"is_reverse": true, "node_id": "13764"}}, {"edit": [{"from_length": 17, "to_length": 17}], "position": {"is_reverse": true, "node_id": "7474"}}], "name": "M01853:259:000000000-KCLW9:1:1101:20339:8721"}, "quality": "EiAeICAlJSUlIyUlJiYmJR4kJiYeJCciJycnIicnJiYiJyYnIiciJyMhJSYmJg8dJA4dHyEiEB0dDiQkDg8dJA4SEh4RDx4kDg8eJREPDg4hIw4OIQ4RDw8bEB4PHQwMGRkNGQ8PDQwMDAweDBgMDAwMGBgaDQ0NDQ0MGBghDAwNDQ0YDRofDA0NDA0YDhgODhgODg4O", "sample_name": "-p", "score": 106, "sequence": "CTGGTTAAGATAAAAGGGGGTACCATGGTCGATGAACAAGTGGCGGCGGAACACGGCACCGTCAGCCACACAAACAGCCGGGAGGAAGATGGCGGGGTCCACGAGCGGCGTGGCTTGGCCACGGGGAGAGGGAGGGAGGCTTTTATAAGG", "time_used": 0.00017778100000000001}
```

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

```
Place stacktrace here.
```

**5. What data and command can the vg dev team use to make the problem happen?**

**6. What does running `vg version` say?**

```
vg version v1.49.0 "Peschici"
Compiled with g++ (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0 on Linux
Linked against libstd++ 20220421
Built by anovak@octagon
```

**Comment by glennhickey** [comment]
Doesn't look familiar but Github Blame doesn't lie: seems to be on me.  @wangl3 can you please share `ebv.pg` and `mapped.gam` and I'll take a look. 

But, I don't see the application for what you are doing.  I'd advice surjecting your original GAM to BAM, and then using a linear variatn caller (DeepVariant, FreeBayes etc.) to analyse the variants, rather than going through augment.

**Comment by wangl3** [comment]
@adamnovak @glennhickey 
Thanks for your comments. There's no issue when I surject the original GAM to BAM. However only 571 variants were called from this BAM using DeepVariant, which is even less than the graph calls from mapped.gam (1262 variants). So I want to try augmented graph to include new variants. 2238 variants were called from augmented graph. As recommended in your README page, I tried to convert this augmented graph to BAM and followed by deepvariant call, in order to see if I can get more variants. But I encountered errors as described above.  
I am wondering if I misunderstand what augmentation process does. or is it because I am working on the virus strains and not sure whether there is special optimization needed. Looking forward to your reply! Thanks.

```
# graph called 1262 variants
~/tools/vg pack -x ebv.d2.xg -g mapped.gam -Q 5 -s 5 -o mapped.pack
~/tools/vg call ebv.d2.xg -k mapped.pack -a > graph_calls.vcf

# Deepvariant called 571 variants 
~/tools/vg surject -t 4 -b -x ebv.d2.xg mapped.gam > ebv_mapped.bam &
samtools sort -@ 4 ebv_mapped.bam -o ebv_mapped_sorted.bam &
samtools index -@ 4 ebv_mapped_sorted.bam &
samtools view ebv_mapped_sorted.bam -h |sed -e 's/NC007605#0#//g' |samtools view -b - > ebv_mapped_sorted_modref.bam
gatk MarkDuplicates -I ebv_mapped_sorted_modref.bam -O ebv_mapped_sorted_modref.removedup.bam -M ebv_mapped_sorted_modref.metrics.txt
samtools index -@ 4 ebv_mapped_sorted_modref.removedup.bam

# graph called 2238 variants
~/tools/vg call ebv-aug.pg -r ebv-aug.snarls -k ebv-aug.pack -s CHS178 > ebv-aug-calls.vcf
```

**Comment by adamnovak** [comment]
I don't think this is a virus-specific problem; we just never tried to use these tools in this combination before.

What augmenting does, is to take candidate variants from your reads, where they are e.g. supported by multiple reads, and add them to the graph so that `vg call`, which only thinks about variants in a graph, can call them. It also edits the reads for you, to express them in terms of the new graph (which has different node numbers and node sequences than the old graph, since some nodes were cut).

As part of the process of rewriting the reads in terms of the new graph's nodes, it looks like we also try to write the variant candidates that we *didn't* put in the augmented graph (because e.g. only one read supported them) out of the reads. At least for insertions and deletions. I think the net result is supposed to be that the rewritten reads only visit nodes and edges that are in the augmented graph, and never skip or add bases. I think we do this because that makes processing them easier in `vg call`.

I'm not actually sure that I would expect using a surjected version of the rewritten reads with DeepVariant to produce more variants than the surjected version of the original reads. Surjection just outputs a mapping of the read to the linear reference, based on how they are both embedded in the graph. Augmenting the graph and rewriting the read isn't really going to move the read around relative to the linear reference, and I don't know that it will usefully clean up the base-level mappings around variants (like e.g. indel realignment is meant to). If anything, I would expect fewer variants called, because we are rewriting some variant candidates out of the reads.

To the extent it does anything, it might confuse the DeepVariant model, because we definitely haven't trained one on reads mapped to augmented graphs.

Are you actually dealing with a diploid (or haploid clonal) set of virus reads to call variants in? Both `vg call` and DeepVariant are going to assume that you mean to be looking at a diploid sample and will refrain from calling variants that might be in 10% of your viral reads. If you just want a survey of all the possible variation in a population of viruses, you might want to use `vg augment` to get a graph that features everything even remotely likely from the reads, and then run that graph through `vg deconstruct` to get a VCF describing all the candidate variants from the graph, without any genotype calls.
